# МЛ Практика 1 Загрядский БПИ-23-1


# Набор данных согласно варианту

Мой вариан - 5, поэтому будем работать с базой, которая содержит информацию о специалистах из разных отраслей


5. https://www.kaggle.com/datasets/navjotkaushal/employees-dataset



## Описание:
В этом разделе вы загрузите набор данных wine.csv, ознакомитесь с его структурой и проведёте первичный обзор.

In [ ]:
# установка зависимостей
!pip install pandas matplotlib seaborn

In [ ]:
# импортируем библиотеки
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Описание:
В этом разделе вы загрузите набор данных wine.csv, ознакомитесь с его структурой и проведёте первичный обзор.

In [ ]:
PATH_TO_FILE: str = "/Employeers_data.csv" # Загрузка набора данных

df = pd.read_csv(PATH_TO_FILE) # функция превращение данных из CSV в DataFrame

In [ ]:
# вывод датафрейма
df # p.s. обычно переменную с набором данных называют df, придерживайтесь этого
# наименования в дальнейшем

id  age     Sex            Occupation              Field  \
0         1   70  Female          Intermediate                 CA   
1         2   41    Male  Working professional                 CA   
2         3   33  Female          Intermediate            Finance   
3         4   50    Male           Experienced                 CA   
4         5   47  Female  Working professional                 CA   
...     ...  ...     ...                   ...                ...   
9995   9996   43  Female           Experienced            Finance   
9996   9997   21  Female          Intermediate                 IT   
9997   9998   62    Male          Intermediate  Digital marketing   
9998   9999   70    Male               Fresher                 IT   
9999  10000   53    Male          Intermediate  Digital marketing   

     Working Hours Package   Job Type                         Company Type  \
0             1 Hr   6 LPA     Remote                     Non-Profit / NGO   
1             3 Hr  27 LPA     Remote                     Consultancy Firm   
2             7 Hr  30 LPA     Remote  Outsourcing / Service-based Company   
3             4 Hr  14 LPA     Hybrid      Contract / Temporary Employment   
4             4 Hr  17 LPA     Hybrid          Small Business / Local Firm   
...            ...     ...        ...                                  ...   
9995          6 Hr   9 LPA     Hybrid  Big MNC (Multinational Corporation)   
9996          5 Hr  12 LPA     Hybrid                Startup (early-stage)   
9997          1 Hr  21 LPA     Remote                     Large Enterprise   
9998         10 Hr  29 LPA  Full-Time                     Non-Profit / NGO   
9999          2 Hr  22 LPA     Remote          Small Business / Local Firm   

     Marital Status Location                          Cerifications  \
0          Divorced    Metro      CFA (Chartered Financial Analyst)   
1          Divorced    Rural      AWS Certified Solutions Architect   
2            Single     City               Certified Data Scientist   
3           Married    Rural       TensorFlow Developer Certificate   
4            Single     City      CFA (Chartered Financial Analyst)   
...             ...      ...                                    ...   
9995       Divorced     City                         Lean Six Sigma   
9996       Divorced     City                             Cisco CCNA   
9997       Divorced    Rural                             Cisco CCNA   
9998        Married    Metro  PMP (Project Management Professional)   
9999        Married     City                      CompTIA Security+   

                 Education_level              Bonus           Allowances  \
0     Professional Certification   ₹5,000 - ₹20,000    Medical Allowance   
1     Professional Certification           No Bonus  Multiple Allowances   
2                        Diploma           No Bonus    Medical Allowance   
3              Bachelor's Degree           No Bonus   Internet Allowance   
4                    High School  ₹20,000 - ₹50,000  Multiple Allowances   
...                          ...                ...                  ...   
9995           Bachelor's Degree  ₹20,000 - ₹50,000          Company Car   
9996                     Diploma   ₹5,000 - ₹20,000                  NaN   
9997           Bachelor's Degree           No Bonus    Housing Allowance   
9998                     Diploma           No Bonus    Housing Allowance   
9999  Professional Certification           No Bonus   Internet Allowance   

                         OverTime Pay Salary Per Month  
0                           ₹200/hour         77000 Rs  
1                           ₹300/hour         41000 Rs  
2     Compensatory Off (No extra pay)         35000 Rs  
3                         No Overtime         95000 Rs  
4                           ₹300/hour        145000 Rs  
...                               ...              ...  
9995                        ₹100/hour         67000 Rs  
9996  Compensatory Of

In [ ]:
# Вывод первых 5 строк
df.head()

# вывод 5 последних строк
#df.tail()

id  age     Sex            Occupation    Field Working Hours Package  \
0   1   70  Female          Intermediate       CA          1 Hr   6 LPA   
1   2   41    Male  Working professional       CA          3 Hr  27 LPA   
2   3   33  Female          Intermediate  Finance          7 Hr  30 LPA   
3   4   50    Male           Experienced       CA          4 Hr  14 LPA   
4   5   47  Female  Working professional       CA          4 Hr  17 LPA   

  Job Type                         Company Type Marital Status Location  \
0   Remote                     Non-Profit / NGO       Divorced    Metro   
1   Remote                     Consultancy Firm       Divorced    Rural   
2   Remote  Outsourcing / Service-based Company         Single     City   
3   Hybrid      Contract / Temporary Employment        Married    Rural   
4   Hybrid          Small Business / Local Firm         Single     City   

                       Cerifications             Education_level  \
0  CFA (Chartered Financial Analyst)  Professional Certification   
1  AWS Certified Solutions Architect  Professional Certification   
2           Certified Data Scientist                     Diploma   
3   TensorFlow Developer Certificate           Bachelor's Degree   
4  CFA (Chartered Financial Analyst)                 High School   

               Bonus           Allowances                     OverTime Pay  \
0   ₹5,000 - ₹20,000    Medical Allowance                        ₹200/hour   
1           No Bonus  Multiple Allowances                        ₹300/hour   
2           No Bonus    Medical Allowance  Compensatory Off (No extra pay)   
3           No Bonus   Internet Allowance                      No Overtime   
4  ₹20,000 - ₹50,000  Multiple Allowances                        ₹300/hour   

  Salary Per Month  
0         77000 Rs  
1         41000 Rs  
2         35000 Rs  
3         95000 Rs  
4        145000 Rs

In [ ]:
# Информация о наборе данных
df.info() # инфо о датасете название колонки, сколько не пропущенных, дефолтный тип данных

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   id                10000 non-null  int64 
 1   age               10000 non-null  int64 
 2   Sex               10000 non-null  object
 3   Occupation        10000 non-null  object
 4   Field             10000 non-null  object
 5   Working Hours     10000 non-null  object
 6   Package           10000 non-null  object
 7   Job Type          10000 non-null  object
 8   Company Type      10000 non-null  object
 9   Marital Status    10000 non-null  object
 10  Location          10000 non-null  object
 11  Cerifications     10000 non-null  object
 12  Education_level   10000 non-null  object
 13  Bonus             10000 non-null  object
 14  Allowances        8672 non-null   object
 15  OverTime Pay      10000 non-null  object
 16  Salary Per Month  10000 non-null  object
dtypes: int64(2), 

In [ ]:
# Статистический разбор
df.describe() # include=all (вывод включая категориальные)

id           age
count  10000.00000  10000.000000
mean    5000.50000     43.680400
std     2886.89568     15.268692
min        1.00000     18.000000
25%     2500.75000     31.000000
50%     5000.50000     44.000000
75%     7500.25000     57.000000
max    10000.00000     70.000000

In [ ]:
# преобразование в категориальные
# df.columnName = df.columnName.astype('category')
# or
# df['column Name'] = df['column Name'].astype('category')

In [ ]:
# Вывод колонок в наборе данных
df.columns

Index(['id', 'age', 'Sex', 'Occupation', 'Field', 'Working Hours', 'Package',
       'Job Type', 'Company Type', 'Marital Status', 'Location',
       'Cerifications', 'Education_level', 'Bonus', 'Allowances',
       'OverTime Pay', 'Salary Per Month'],
      dtype='object')

## Определение типов переменных

In [ ]:
# Список столбцов и их типов данных
print(df.dtypes)

id                   int64
age                  int64
Sex                 object
Occupation          object
Field               object
Working Hours       object
Package             object
Job Type            object
Company Type        object
Marital Status      object
Location            object
Cerifications       object
Education_level     object
Bonus               object
Allowances          object
OverTime Pay        object
Salary Per Month    object
dtype: object


- Числовые переменные обычно имеют тип int64 или float64.
- Категориальные переменные могут иметь тип object или int (например метки класса)

## Статистический разбор данных

1. **Посчитайте количество уникальных значений для каждой переменной:**

**⏰ задача**

Определить типы переменных
- номинальные переменные (не имеют порядка и сравнения)
- порядковые пиеременные (имеют порядок, но не имеют смысла мат операции)
- числовые непрерывные переменные (имеют порядок, есть смысл мат операций)

Вывести статистику для числовых и для категоиральных перменных раздельно.

p.s. если существует несколько гипотез к какому типу относительно одна и та же переменной - объяснить почему

In [ ]:
# Количество уникальных значений
for column in df.columns:
    print(f'{column}: {df[column].nunique()} уникальных значений')

id: 10000 уникальных значений
age: 53 уникальных значений
Sex: 2 уникальных значений
Occupation: 4 уникальных значений
Field: 4 уникальных значений
Working Hours: 10 уникальных значений
Package: 26 уникальных значений
Job Type: 3 уникальных значений
Company Type: 14 уникальных значений
Marital Status: 3 уникальных значений
Location: 3 уникальных значений
Cerifications: 11 уникальных значений
Education_level: 6 уникальных значений
Bonus: 3 уникальных значений
Allowances: 7 уникальных значений
OverTime Pay: 6 уникальных значений
Salary Per Month: 63 уникальных значений


# **Номинальные (категориальные без порядка):**

id - идентификатор (уникальный для каждого наблюдения)

Sex - пол (Male/Female)

Field - сфера деятельности (CA, IT, Finance, Digital marketing)

Job Type - тип работы (Remote, Hybrid, Full-Time)

Company Type - тип компании (14 категорий)

Marital Status - семейное положение (Single, Married, Divorced)

Location - местоположение (Metro, City, Rural)

Cerifications - сертификации (11 категорий)

Allowances - надбавки (7 категорий)

OverTime Pay - оплата сверхурочных (6 категорий)

# **Порядковые (категориальные с порядком):**

Occupation - уровень опыта (Fresher, Intermediate, Experienced, Working professional) - имеет явную иерархию

Education_level - уровень образования (High School, Diploma, Bachelor's Degree, Master's Degree, PhD, Professional Certification) - имеет четкий порядок

Bonus - бонусы (No Bonus, "₹5,000 - ₹20,000", "₹20,000 - ₹50,000") - имеет порядок по сумме

Working Hours - рабочие часы (от 1 Hr до 10 Hr) - имеет порядок, но интервалы не равны

# **Числовые непрерывные:**

age - возраст (18-70 лет)

Package - пакет оплаты (5-30 LPA) - годовой доход

Salary Per Month - месячная зарплата (25000-149000 Rs)

# **Статистика для числовых переменных**

In [ ]:
# Для переменных age, Package, Salary Per Month
numerical_stats = df[['age', 'Package', 'Salary Per Month']].describe()
print(numerical_stats)

                age
count  10000.000000
mean      43.680400
std       15.268692
min       18.000000
25%       31.000000
50%       44.000000
75%       57.000000
max       70.000000


# **Статистика для категориальных переменных**


In [ ]:
# Для номинальных и порядковых переменных
categorical_columns = ['Sex', 'Occupation', 'Field', 'Working Hours', 'Job Type',
                      'Company Type', 'Marital Status', 'Location', 'Cerifications',
                      'Education_level', 'Bonus', 'Allowances', 'OverTime Pay']

for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts())


--- Sex ---
Sex
Male      5001
Female    4999
Name: count, dtype: int64

--- Occupation ---
Occupation
Experienced             2570
Intermediate            2491
Working professional    2491
Fresher                 2448
Name: count, dtype: int64

--- Field ---
Field
IT                   2588
CA                   2503
Digital marketing    2500
Finance              2409
Name: count, dtype: int64

--- Working Hours ---
Working Hours
3     1029
6     1027
7     1017
10    1013
1     1008
2     1006
9      992
4      987
8      973
5      948
Name: count, dtype: int64

--- Job Type ---
Job Type
Full-Time    3408
Hybrid       3307
Remote       3285
Name: count, dtype: int64

--- Company Type ---
Company Type
Contract / Temporary Employment                                               767
Product-based Company                                                         762
Mid-sized Company                                                             747
Non-Profit / NGO                          

# **Гипотезы по спорным переменным**


Working Hours - можно рассматривать как числовую, если преобразовать "1 Hr" в 1, "2 Hr" в 2 и т.д.



Bonus - диапазоны значений ("₹5,000 - ₹20,000") затрудняют точные математические операции, поэтому лучше рассматривать как порядковую.

## **4. Визуализация распределения переменных**

### **Описание:**
Создайте визуализации для понимания распределения числовых и категориальных переменных.

### **Шаги:**


In [1]:
# Гистограммы для числовых переменных
plt.figure(figsize=(16, 12))

# Salary Per Month
plt.subplot(3, 2, 1)
sns.histplot(df['Salary Per Month'], kde=True)
plt.title('Распределение месячной зарплаты')

# Working Hours
plt.subplot(3, 2, 2)
sns.histplot(df['Working Hours'], kde=True)
plt.title('Распределение рабочих часов')

# Package
plt.subplot(3, 2, 3)
sns.histplot(df['Package'], kde=True)
plt.title('Распределение годового пакета')

# Age
plt.subplot(3, 2, 4)
sns.histplot(df['age'], kde=True)
plt.title('Распределение возраста')

# OverTime Pay
plt.subplot(3, 2, 5)
sns.histplot(df['OverTime Pay'], kde=True)
plt.title('Распределение оплаты сверхурочных')

plt.tight_layout()
plt.show()

NameError: name 'plt' is not defined


1. **Зарплата в месяц**
* График показывает, что большинство людей получают примерно **от 80 до 120 тысяч**
* Форма графика похожа на колокол (как нормальное распределение)
* Немного больше людей получают зарплату ближе к нижней границе

2. **Часы работы**
* Люди в основном работают **от 2 до 10 часов** в день
* График ровный, без резких скачков
* Видно, что распределение почти равномерное

3. **Годовой доход**
* Большинство получает **от 10 до 20** в год
* График тоже похож на колокол
* Значения распределены довольно плотно

4. **Возраст сотрудников**
* Самые частые значения - **от 30 до 50 лет**
* Это нормальное распределение, как и у зарплаты
* Всё довольно равномерно

5. **Доплата за сверхурочную работу**
* Основная масса получает **от 1500 до 2000**
* График опять похож на колокол
* Нет больших выбросов

Если посмотреть на категории (например, мужчины/женщины), можно заметить:
* Зарплаты у мужчин и женщин примерно одинаковые
* Часы работы тоже не сильно отличаются
* По возрасту группы распределены равномерно

В общем, графики показывают, что:
* Большинство показателей распределены нормально
* Значения не сильно разбросаны
* Нет каких-то диких отклонений

2. **Столбчатая диаграмма для категориальной переменной**

In [2]:

# Создаем столбчатую диаграмму
plt.figure(figsize=(8, 6))
sns.countplot(x='Sex', data=df,
              palette='pastel',
              order=df['Sex'].value_counts().index)

# Добавляем аннотации
for p in plt.gca().patches:
    plt.gca().annotate(f'{(p.get_height())}',
                      (p.get_x() + p.get_width() / 2., p.get_height()),
                      ha='center', va='center', fontsize=10, color='black',
                      xytext=(0, 5), textcoords='offset points')

plt.title('Распределение по полу', fontsize=14)
plt.xlabel('Пол', fontsize=12)
plt.ylabel('Количество наблюдений', fontsize=12)
plt.grid(axis='y', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

NameError: name 'plt' is not defined

На графике представлено распределение по полу. Видно, что количество наблюдений для мужчин и женщин практически одинаково: 5001 для мужчин и 4999 для женщин. Это указывает на сбалансированное распределение по полу в выборке.

Если добавить категориальную переменную с помощью параметра hue, можно проанализировать, как распределение по полу связано с другими факторами.

In [ ]:
plt.figure(figsize=(12, 6))
sns.barplot(
    x='Sex',
    y='Salary Per Month',
    hue='Occupation',
    data=df,
    ci=95  # доверительный интервал
)

plt.title('Распределение зарплат по полу и уровню занятости')
plt.xlabel('Пол')
plt.ylabel('Средняя зарплата (рупии)')
plt.legend(title='Уровень занятости')
plt.grid(True)
plt.show()

**Боксплот для числовой переменной:**

In [ ]:

# Создаем боксплот для зарплаты
plt.figure(figsize=(12, 8))
boxplot = sns.boxplot(
    y='Salary Per Month',
    data=df,
    palette='viridis',
    showfliers=True  # показывать выбросы
)

# Добавляем медиану красным цветом
sns.pointplot(
    y='Salary Per Month',
    data=df,
    color='red',
    ci=None
)

plt.title('Распределение зарплат сотрудников', fontsize=16)
plt.ylabel('Месячная зарплата (рупии)', fontsize=14)
plt.xlabel('')
plt.grid(True, linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

Выбросы: На графике виден один выброс, отмеченный красной точкой, который находится значительно ниже основного диапазона зарплат. Это может указывать на аномально низкую зарплату, которая требует дополнительного анализа.

Медиана: Медиана зарплат находится примерно на уровне 85 000 рупий, что является центральной точкой распределения. Это значение делит данные на две равные части: половина сотрудников получает зарплату выше этой суммы, а другая половина — ниже.

Разброс данных: Межквартильный размах (IQR) показывает, что 50% сотрудников имеют зарплаты в диапазоне от около 80 000 до 100 000 рупий. Это говорит о том, что зарплаты в целом распределены достаточно равномерно в этом интервале.

In [3]:

# Выберем только числовые столбцы
numeric_cols = df.select_dtypes(include=['float64', 'int64']).columns

# Проверим, какие столбцы выбраны
print("Числовые столбцы:", numeric_cols)

# Создадим корреляционную матрицу
corr_matrix = df[numeric_cols].corr()

# Визуализация тепловой карты
plt.figure(figsize=(14, 10))
sns.heatmap(
    corr_matrix,
    annot=True,  # показывать значения корреляции
    cmap='coolwarm',  # цветовая схема
    fmt=".2f",  # формат чисел
    vmin=-1,  # минимальное значение
    vmax=1,  # максимальное значение
    square=True,  # квадратные ячейки
    linewidths=0.5  # толщина линий между ячейками
)
plt.title('Корреляционная матрица числовых переменных', fontsize=16)
plt.xticks(rotation=45)
plt.yticks(rotation=0)
plt.tight_layout()
plt.show()

NameError: name 'df' is not defined



1. **Переменные, которые не коррелируют (коэффициент близок к 0):**
   id с Working Hours, Package, Salary Per Month.
   age с Working Hours, Package, Salary Per Month.
   Working Hours с Package, Salary Per Month.
   Package с Salary Per Month.

2. **Переменные с сильной положительной корреляцией (коэффициент ≥ 0.7):**
   id с id (1.00).
   age с age (1.00).
   Working Hours с Working Hours (1.00).
   Package с Package (1.00).
   Salary Per Month` с Salary Per Month (1.00).

3. **Переменные с слабой корреляцией (коэффициент < 0.3):**
   - Все пары переменных, кроме тех, где коэффициент равен 1.00.

